### ISSUE TITLE
Expanded documentation for TypeScript 😈

### ISSUE THREAD

**Issue by atrauzzi** [dcodeIO/protobuf.js#666]
protobuf.js version: 6.6.2

This is more of a documentation issue..  The documentation in `README.md` doesn't really give me a working understanding of what my workflows need to be with this library.

If I'm authoring code in TypeScript and wish to make use of this library, what do I have to set up and what commands or `proto` files do I have to leverage to get everything to a point where I can go back to using domain classes?

Again, it's a documentation thing, but if I'm not entirely clear on what I have to do, what generated code I need, where to put it or how to leverage it, perhaps I'm not alone? :)

**Comment by dcodeIO** [comment]
Copying the already slightly modified TypeScript paragraph here so we can work on it:

### Usage with TypeScript

The library ships with its own [type definitions](https://github.com/dcodeIO/protobuf.js/blob/master/index.d.ts) and modern editors like [Visual Studio Code](https://code.visualstudio.com/) should automatically detect and use them for code completion when following this pattern:

```ts
import * as protobuf from "protobufjs";
import * as Long from "long"; // optional

...
```

TODO

See also: [Generating your own TypeScript definitions](https://github.com/dcodeIO/protobuf.js#generating-typescript-definitions-from-static-modules)

**Comment by dcodeIO** [comment]
Let me know if this leaves any open questions.

**Comment by atrauzzi** [comment]
@dcodeIO - Awesome, yeah -- so let's see if I can get all of this out in one go.  Sometimes it's tough to know all problems in advance ;)

 - Is there a way to generate code to a single file per message type? My thought there is that I'm trying to leverage protocol buffers for serialization of my domain models but would still like to be able to have behaviour in the types.
 - As a followup to that, if I deserialize data with protobufjs, will I get instances of my models, or just JS objects that match the `.d.ts` file(s)?
 - Alternatively, Is there any way to use this library by having protobufjs attach/wrap my types without having to use generated code or extend a type?

Apologies for the _"Hi I'm new here"_-ness of my questions.  Once I know the right way to use this library, things will click.

**Comment by dcodeIO** [comment]
> Is there a way to generate code to a single file per message type?

Not yet, unfortunately. Some day.

> if I deserialize data with protobufjs, will I get instances of my models

> Is there any way to use this library by having protobufjs attach/wrap my types without having to use generated code or extend a type

With reflection, you'll get runtime message instances, and you can [extend the respective class](https://github.com/dcodeIO/protobuf.js#using-custom-classes) with any custom functionality.

**Comment by atrauzzi** [comment]
Great, I think this gives me a loose idea of what I can do right away, with some decent wishlist candidates.

Last round of questions (I think):

 - For generated code, any best practices or suggestions on how to manage it cleanly?
  - If I'm working in a TS project, where should I put my generated `.js` files and the corresponding `.d.ts`?
  - What's your take on committing the generated code to source control? Normally I don't like to put anything that isn't authored by a person in source control, but I don't know if this qualifies as an exception given that my actual domain classes will be depending on them.

**Comment by dcodeIO** [comment]
Well, that's completely up to you, but that's what I'd do:

> If I'm working in a TS project, where should I put my generated .js files and the corresponding .d.ts?

* Single bundle containing all the definitions? Put it next to the code relying on it.
* Multiple bundles? Put everything in a `protobuf` folder.

> What's your take on committing the generated code to source control?

* Generated static code which is manually edited? Put it in source control, of course.
* Generated static code which is not manually edited? Use a JSON module with a d.ts generated for the static counterpart instead, because JSON is pretty much conflict free when it comes to linters etc. and can be easily extended with additional fields or even types manually without re-generating anything.

**Comment by atrauzzi** [comment]
I'm not sure I follow how _"reflection"_ is being used in this specific scenario.  Normally I understand it to be when you query and modify the type system at runtime - which JS/TS lack.  What I'm seeing here is a fluent API to build metadata, rather than using JSON or protobuf files.

One or more of what you've described so far might be what I'm after here, but I'm having a tough time aligning it to my understanding or picking out the correct approaches.

So let's take a new scenario: Assume I don't want to use generated code, I have a bunch of models I've created in TS.  Is there a way to load the metadata (whether through the fluent API, protobuf or JSON) and then use protobufjs as an infrastructure-service in my application?

**Comment by dcodeIO** [comment]
I've added additional information on reflection-based and code-based use cases to the README.

Basically, everything which is not static code is reflection based.

**Comment by atrauzzi** [comment]
Alright!  Looks like I've got things more or less *exactly* how I'd like!  The `es6` combined with the `json-module` output for `pbjs` was exactly what I was after.  I pass in the `proto` files, it generates everything for me and I can pull it all in where necessary.

What an amazing job you've done here.

I've got one minor suggestion to appease my linter which I'll open as another ticket.  I don't know if you'd like to ask me any specific questions to help with further improving the documentation.  But otherwise, feel free to close!
